// @wordpress/format-library/build/link/modal.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _components = require("@wordpress/components");

var _screens = _interopRequireDefault(require("./modal-screens/screens"));

var _linkSettingsScreen = _interopRequireDefault(require("./modal-screens/link-settings-screen"));

var _linkPickerScreen = _interopRequireDefault(require("./modal-screens/link-picker-screen"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
const ModalLinkUI = _ref => {
  let {
    isVisible,
    ...restProps
  } = _ref;
  return (0, _element.useMemo)(() => {
    return (0, _element.createElement)(_components.BottomSheet, {
      isVisible: isVisible,
      hideHeader: true,
      onClose: restProps.onClose,
      hasNavigation: true
    }, (0, _element.createElement)(_components.BottomSheet.NavigationContainer, {
      animate: true,
      main: true
    }, (0, _element.createElement)(_components.BottomSheet.NavigationScreen, {
      name: _screens.default.settings
    }, (0, _element.createElement)(_linkSettingsScreen.default, (0, _extends2.default)({
      isVisible: isVisible
    }, restProps))), (0, _element.createElement)(_components.BottomSheet.NavigationScreen, {
      name: _screens.default.picker,
      isScrollable: true,
      fullScreen: true
    }, (0, _element.createElement)(_linkPickerScreen.default, null))));
  }, [isVisible]);
};

var _default = (0, _components.withSpokenMessages)(ModalLinkUI);

exports.default = _default;
