// @wordpress/format-library/build/superscript/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.superscript = void 0;

var _element = require("@wordpress/element");

var _i18n = require("@wordpress/i18n");

var _richText = require("@wordpress/rich-text");

var _blockEditor = require("@wordpress/block-editor");

var _icons = require("@wordpress/icons");

/**
 * WordPress dependencies
 */
const name = 'core/superscript';
const title = (0, _i18n.__)('Superscript');
const superscript = {
  name,
  title,
  tagName: 'sup',
  className: null,

  edit(_ref) {
    let {
      isActive,
      value,
      onChange,
      onFocus
    } = _ref;

    function onToggle() {
      onChange((0, _richText.toggleFormat)(value, {
        type: name,
        title
      }));
    }

    function onClick() {
      onToggle();
      onFocus();
    }

    return (0, _element.createElement)(_blockEditor.RichTextToolbarButton, {
      icon: _icons.superscript,
      title: title,
      onClick: onClick,
      isActive: isActive,
      role: "menuitemcheckbox"
    });
  }

};
exports.superscript = superscript;
