// @wordpress/format-library/build-module/code/index.js
import { createElement, Fragment } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { toggleFormat, remove, applyFormat } from '@wordpress/rich-text';
import { RichTextToolbarButton, RichTextShortcut } from '@wordpress/block-editor';
import { code as codeIcon } from '@wordpress/icons';
const name = 'core/code';

const title = __('Inline code');

export const code = {
  name,
  title,
  tagName: 'code',
  className: null,

  __unstableInputRule(value) {
    const BACKTICK = '`';
    const {
      start,
      text
    } = value;
    const characterBefore = text.slice(start - 1, start); // Quick check the text for the necessary character.

    if (characterBefore !== BACKTICK) {
      return value;
    }

    const textBefore = text.slice(0, start - 1);
    const indexBefore = textBefore.lastIndexOf(BACKTICK);

    if (indexBefore === -1) {
      return value;
    }

    const startIndex = indexBefore;
    const endIndex = start - 2;

    if (startIndex === endIndex) {
      return value;
    }

    value = remove(value, startIndex, startIndex + 1);
    value = remove(value, endIndex, endIndex + 1);
    value = applyFormat(value, {
      type: name
    }, startIndex, endIndex);
    return value;
  },

  edit(_ref) {
    let {
      value,
      onChange,
      onFocus,
      isActive
    } = _ref;

    function onClick() {
      onChange(toggleFormat(value, {
        type: name,
        title
      }));
      onFocus();
    }

    return createElement(Fragment, null, createElement(RichTextShortcut, {
      type: "access",
      character: "x",
      onUse: onClick
    }), createElement(RichTextToolbarButton, {
      icon: codeIcon,
      title: title,
      onClick: onClick,
      isActive: isActive,
      role: "menuitemcheckbox"
    }));
  }

};
