// @wordpress/format-library/build/strikethrough/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.strikethrough = void 0;

var _element = require("@wordpress/element");

var _i18n = require("@wordpress/i18n");

var _richText = require("@wordpress/rich-text");

var _blockEditor = require("@wordpress/block-editor");

var _icons = require("@wordpress/icons");

/**
 * WordPress dependencies
 */
const name = 'core/strikethrough';
const title = (0, _i18n.__)('Strikethrough');
const strikethrough = {
  name,
  title,
  tagName: 's',
  className: null,

  edit(_ref) {
    let {
      isActive,
      value,
      onChange,
      onFocus
    } = _ref;

    function onClick() {
      onChange((0, _richText.toggleFormat)(value, {
        type: name,
        title
      }));
      onFocus();
    }

    return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_blockEditor.RichTextShortcut, {
      type: "access",
      character: "d",
      onUse: onClick
    }), (0, _element.createElement)(_blockEditor.RichTextToolbarButton, {
      icon: _icons.formatStrikethrough,
      title: title,
      onClick: onClick,
      isActive: isActive,
      role: "menuitemcheckbox"
    }));
  }

};
exports.strikethrough = strikethrough;
