// @wordpress/format-library/build/bold/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.bold = void 0;

var _element = require("@wordpress/element");

var _i18n = require("@wordpress/i18n");

var _richText = require("@wordpress/rich-text");

var _blockEditor = require("@wordpress/block-editor");

var _icons = require("@wordpress/icons");

/**
 * WordPress dependencies
 */
const name = 'core/bold';
const title = (0, _i18n.__)('Bold');
const bold = {
  name,
  title,
  tagName: 'strong',
  className: null,

  edit(_ref) {
    let {
      isActive,
      value,
      onChange,
      onFocus
    } = _ref;

    function onToggle() {
      onChange((0, _richText.toggleFormat)(value, {
        type: name,
        title
      }));
    }

    function onClick() {
      onChange((0, _richText.toggleFormat)(value, {
        type: name
      }));
      onFocus();
    }

    return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_blockEditor.RichTextShortcut, {
      type: "primary",
      character: "b",
      onUse: onToggle
    }), (0, _element.createElement)(_blockEditor.RichTextToolbarButton, {
      name: "bold",
      icon: _icons.formatBold,
      title: title,
      onClick: onClick,
      isActive: isActive,
      shortcutType: "primary",
      shortcutCharacter: "b"
    }), (0, _element.createElement)(_blockEditor.__unstableRichTextInputEvent, {
      inputType: "formatBold",
      onInput: onToggle
    }));
  }

};
exports.bold = bold;
