// @wordpress/widgets/build/blocks/widget-group/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = void 0;

var _i18n = require("@wordpress/i18n");

var _blocks = require("@wordpress/blocks");

var _icons = require("@wordpress/icons");

var _edit = _interopRequireDefault(require("./edit"));

var _save = _interopRequireDefault(require("./save"));

var _deprecated = _interopRequireDefault(require("./deprecated"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
const metadata = {
  apiVersion: 2,
  name: "core/widget-group",
  category: "widgets",
  attributes: {
    title: {
      type: "string"
    }
  },
  supports: {
    html: false,
    inserter: true,
    customClassName: true,
    reusable: false
  },
  editorStyle: "wp-block-widget-group-editor",
  style: "wp-block-widget-group"
};
exports.metadata = metadata;
const {
  name
} = metadata;
exports.name = name;
const settings = {
  title: (0, _i18n.__)('Widget Group'),
  description: (0, _i18n.__)('Create a classic widget layout with a title that’s styled by your theme for your widget areas.'),
  icon: _icons.group,
  __experimentalLabel: _ref => {
    let {
      name: label
    } = _ref;
    return label;
  },
  edit: _edit.default,
  save: _save.default,
  transforms: {
    from: [{
      type: 'block',
      isMultiBlock: true,
      blocks: ['*'],

      isMatch(attributes, blocks) {
        // Avoid transforming existing `widget-group` blocks.
        return !blocks.some(block => block.name === 'core/widget-group');
      },

      __experimentalConvert(blocks) {
        // Put the selected blocks inside the new Widget Group's innerBlocks.
        let innerBlocks = [...blocks.map(block => {
          return (0, _blocks.createBlock)(block.name, block.attributes, block.innerBlocks);
        })]; // If the first block is a heading then assume this is intended
        // to be the Widget's "title".

        const firstHeadingBlock = innerBlocks[0].name === 'core/heading' ? innerBlocks[0] : null; // Remove the first heading block as we're copying
        // it's content into the Widget Group's title attribute.

        innerBlocks = innerBlocks.filter(block => block !== firstHeadingBlock);
        return (0, _blocks.createBlock)('core/widget-group', { ...(firstHeadingBlock && {
            title: firstHeadingBlock.attributes.content
          })
        }, innerBlocks);
      }

    }]
  },
  deprecated: _deprecated.default
};
exports.settings = settings;
