// @wordpress/list-reusable-blocks/build/components/import-dropdown/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _compose = require("@wordpress/compose");

var _i18n = require("@wordpress/i18n");

var _components = require("@wordpress/components");

var _importForm = _interopRequireDefault(require("../import-form"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function ImportDropdown(_ref) {
  let {
    onUpload
  } = _ref;
  return (0, _element.createElement)(_components.Dropdown, {
    position: "bottom right",
    contentClassName: "list-reusable-blocks-import-dropdown__content",
    renderToggle: _ref2 => {
      let {
        isOpen,
        onToggle
      } = _ref2;
      return (0, _element.createElement)(_components.Button, {
        "aria-expanded": isOpen,
        onClick: onToggle,
        variant: "primary"
      }, (0, _i18n.__)('Import from JSON'));
    },
    renderContent: _ref3 => {
      let {
        onClose
      } = _ref3;
      return (0, _element.createElement)(_importForm.default, {
        onUpload: (0, _compose.pipe)(onClose, onUpload)
      });
    }
  });
}

var _default = ImportDropdown;
exports.default = _default;
