// @wordpress/format-library/build-module/link/modal.native.js
import _extends from "@babel/runtime/helpers/esm/extends";
import { createElement } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { useMemo } from '@wordpress/element';
import { BottomSheet, withSpokenMessages } from '@wordpress/components';
/**
 * Internal dependencies
 */

import screens from './modal-screens/screens';
import LinkSettingsScreen from './modal-screens/link-settings-screen';
import LinkPickerScreen from './modal-screens/link-picker-screen';

const ModalLinkUI = _ref => {
  let {
    isVisible,
    ...restProps
  } = _ref;
  return useMemo(() => {
    return createElement(BottomSheet, {
      isVisible: isVisible,
      hideHeader: true,
      onClose: restProps.onClose,
      hasNavigation: true
    }, createElement(BottomSheet.NavigationContainer, {
      animate: true,
      main: true
    }, createElement(BottomSheet.NavigationScreen, {
      name: screens.settings
    }, createElement(LinkSettingsScreen, _extends({
      isVisible: isVisible
    }, restProps))), createElement(BottomSheet.NavigationScreen, {
      name: screens.picker,
      isScrollable: true,
      fullScreen: true
    }, createElement(LinkPickerScreen, null))));
  }, [isVisible]);
};

export default withSpokenMessages(ModalLinkUI);
