// @wordpress/format-library/build/unknown/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.unknown = void 0;

var _element = require("@wordpress/element");

var _i18n = require("@wordpress/i18n");

var _richText = require("@wordpress/rich-text");

var _blockEditor = require("@wordpress/block-editor");

var _icons = require("@wordpress/icons");

/**
 * WordPress dependencies
 */
const name = 'core/unknown';
const title = (0, _i18n.__)('Clear Unknown Formatting');
const unknown = {
  name,
  title,
  tagName: '*',
  className: null,

  edit(_ref) {
    let {
      isActive,
      value,
      onChange,
      onFocus
    } = _ref;

    function onClick() {
      onChange((0, _richText.removeFormat)(value, name));
      onFocus();
    }

    const selectedValue = (0, _richText.slice)(value);
    const hasUnknownFormats = selectedValue.formats.some(formats => {
      return formats.some(format => format.type === name);
    });

    if (!isActive && !hasUnknownFormats) {
      return null;
    }

    return (0, _element.createElement)(_blockEditor.RichTextToolbarButton, {
      name: "unknown",
      icon: _icons.help,
      title: title,
      onClick: onClick,
      isActive: true
    });
  }

};
exports.unknown = unknown;
