// @wordpress/editor/build-module/components/editor-help/index.native.js
import _extends from "@babel/runtime/helpers/esm/extends";
import { createElement, Fragment } from "@wordpress/element";

/**
 * External dependencies
 */
import { kebabCase } from 'lodash';
import { SafeAreaView, ScrollView, StyleSheet, View } from 'react-native';
import { TransitionPresets } from '@react-navigation/stack';
/**
 * WordPress dependencies
 */

import { BottomSheet, BottomSheetConsumer, PanelBody } from '@wordpress/components';
import { __ } from '@wordpress/i18n';
import { helpFilled, plusCircleFilled, trash, cog } from '@wordpress/icons';
import { useSelect } from '@wordpress/data';
import { store as editorStore } from '@wordpress/editor';
import { requestContactCustomerSupport, requestGotoCustomerSupportOptions } from '@wordpress/react-native-bridge';
/**
 * Internal dependencies
 */

import styles from './style.scss';
import HelpDetailNavigationScreen from './help-detail-navigation-screen';
import HelpTopicRow from './help-topic-row';
import HelpGetSupportButton from './help-get-support-button';
import IntroToBlocks from './intro-to-blocks';
import AddBlocks from './add-blocks';
import MoveBlocks from './move-blocks';
import RemoveBlocks from './remove-blocks';
import CustomizeBlocks from './customize-blocks';
import moveBlocksIcon from './icon-move-blocks';
import HelpSectionTitle from './help-section-title';
const HELP_TOPICS = [{
  label: __('What is a block?'),
  icon: helpFilled,
  view: createElement(IntroToBlocks, null)
}, {
  label: __('Add blocks'),
  icon: plusCircleFilled,
  view: createElement(AddBlocks, null)
}, {
  label: __('Move blocks'),
  icon: moveBlocksIcon,
  view: createElement(MoveBlocks, null)
}, {
  label: __('Remove blocks'),
  icon: trash,
  view: createElement(RemoveBlocks, null)
}, {
  label: __('Customize blocks'),
  icon: cog,
  view: createElement(CustomizeBlocks, null)
}];

function EditorHelpTopics(_ref) {
  let {
    close,
    isVisible,
    onClose,
    showSupport
  } = _ref;
  const {
    postType
  } = useSelect(select => ({
    postType: select(editorStore).getEditedPostAttribute('type')
  }));
  const title = postType === 'page' ? __('How to edit your page') : __('How to edit your post');
  const supportSection = createElement(Fragment, null, createElement(HelpSectionTitle, null, __('Get support')), createElement(HelpGetSupportButton, {
    title: __('Contact support'),
    onPress: requestContactCustomerSupport
  }), createElement(HelpGetSupportButton, {
    title: __('More support options'),
    onPress: requestGotoCustomerSupportOptions
  }));
  return createElement(BottomSheet, {
    isVisible: isVisible,
    onClose: onClose,
    hideHeader: true,
    hasNavigation: true,
    contentStyle: styles.contentContainer,
    testID: "editor-help-modal"
  }, createElement(SafeAreaView, null, createElement(BottomSheet.NavigationContainer, {
    animate: true,
    main: true,
    style: styles.navigationContainer
  }, createElement(BottomSheet.NavigationScreen, {
    isScrollable: true,
    fullScreen: true,
    name: "help-topics"
  }, createElement(View, {
    style: styles.container
  }, createElement(BottomSheet.NavBar, null, createElement(BottomSheet.NavBar.DismissButton, {
    onPress: close,
    iosText: __('Close')
  }), createElement(BottomSheet.NavBar.Heading, null, title)), createElement(BottomSheetConsumer, null, _ref2 => {
    let {
      listProps
    } = _ref2;
    const contentContainerStyle = StyleSheet.flatten(listProps.contentContainerStyle);
    return createElement(ScrollView, _extends({}, listProps, {
      contentContainerStyle: { ...contentContainerStyle,
        paddingBottom: Math.max(listProps.safeAreaBottomInset, contentContainerStyle.paddingBottom),

        /**
         * Remove margin set via `hideHeader`. Combining a header
         * and navigation in this bottom sheet is at odds with the
         * current `BottomSheet` implementation.
         */
        marginTop: 0
      }
    }), createElement(PanelBody, null, createElement(HelpSectionTitle, null, __('The basics')), HELP_TOPICS.map((_ref3, index) => {
      let {
        label,
        icon
      } = _ref3;
      const labelSlug = kebabCase(label);
      const isLastItem = index === HELP_TOPICS.length - 1;
      return createElement(HelpTopicRow, {
        key: labelSlug,
        label: label,
        icon: icon,
        screenName: labelSlug,
        isLastItem: isLastItem
      });
    }), showSupport && supportSection));
  }))), HELP_TOPICS.map(_ref4 => {
    let {
      view,
      label
    } = _ref4;
    const labelSlug = kebabCase(label);
    return createElement(HelpDetailNavigationScreen, {
      key: labelSlug,
      name: labelSlug,
      content: view,
      label: label,
      options: {
        gestureEnabled: false,
        ...TransitionPresets.DefaultTransition
      }
    });
  }))));
}

export default EditorHelpTopics;
