// @wordpress/widgets/build-module/components/move-to-widget-area/index.js
import { createElement } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { DropdownMenu, MenuGroup, MenuItemsChoice, ToolbarGroup, ToolbarItem } from '@wordpress/components';
import { __ } from '@wordpress/i18n';
import { moveTo } from '@wordpress/icons';
export default function MoveToWidgetArea(_ref) {
  let {
    currentWidgetAreaId,
    widgetAreas,
    onSelect
  } = _ref;
  return createElement(ToolbarGroup, null, createElement(ToolbarItem, null, toggleProps => createElement(DropdownMenu, {
    icon: moveTo,
    label: __('Move to widget area'),
    toggleProps: toggleProps
  }, _ref2 => {
    let {
      onClose
    } = _ref2;
    return createElement(MenuGroup, {
      label: __('Move to')
    }, createElement(MenuItemsChoice, {
      choices: widgetAreas.map(widgetArea => ({
        value: widgetArea.id,
        label: widgetArea.name,
        info: widgetArea.description
      })),
      value: currentWidgetAreaId,
      onSelect: value => {
        onSelect(value);
        onClose();
      }
    }));
  })));
}
