// @wordpress/interface/build/components/preferences-modal-tabs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PreferencesModalTabs;

var _element = require("@wordpress/element");

var _compose = require("@wordpress/compose");

var _components = require("@wordpress/components");

var _icons = require("@wordpress/icons");

var _i18n = require("@wordpress/i18n");

/**
 * WordPress dependencies
 */
const PREFERENCES_MENU = 'preferences-menu';

function PreferencesModalTabs(_ref) {
  let {
    sections
  } = _ref;
  const isLargeViewport = (0, _compose.useViewportMatch)('medium'); // This is also used to sync the two different rendered components
  // between small and large viewports.

  const [activeMenu, setActiveMenu] = (0, _element.useState)(PREFERENCES_MENU);
  /**
   * Create helper objects from `sections` for easier data handling.
   * `tabs` is used for creating the `TabPanel` and `sectionsContentMap`
   * is used for easier access to active tab's content.
   */

  const {
    tabs,
    sectionsContentMap
  } = (0, _element.useMemo)(() => {
    let mappedTabs = {
      tabs: [],
      sectionsContentMap: {}
    };

    if (sections.length) {
      mappedTabs = sections.reduce((accumulator, _ref2) => {
        let {
          name,
          tabLabel: title,
          content
        } = _ref2;
        accumulator.tabs.push({
          name,
          title
        });
        accumulator.sectionsContentMap[name] = content;
        return accumulator;
      }, {
        tabs: [],
        sectionsContentMap: {}
      });
    }

    return mappedTabs;
  }, [sections]);
  const getCurrentTab = (0, _element.useCallback)(tab => sectionsContentMap[tab.name] || null, [sectionsContentMap]);
  let modalContent; // We render different components based on the viewport size.

  if (isLargeViewport) {
    modalContent = (0, _element.createElement)(_components.TabPanel, {
      className: "interface-preferences__tabs",
      tabs: tabs,
      initialTabName: activeMenu !== PREFERENCES_MENU ? activeMenu : undefined,
      onSelect: setActiveMenu,
      orientation: "vertical"
    }, getCurrentTab);
  } else {
    modalContent = (0, _element.createElement)(_components.__experimentalNavigatorProvider, {
      initialPath: "/",
      className: "interface-preferences__provider"
    }, (0, _element.createElement)(_components.__experimentalNavigatorScreen, {
      path: "/"
    }, (0, _element.createElement)(_components.Card, {
      isBorderless: true,
      size: "small"
    }, (0, _element.createElement)(_components.CardBody, null, (0, _element.createElement)(_components.__experimentalItemGroup, null, tabs.map(tab => {
      return (0, _element.createElement)(_components.__experimentalNavigatorButton, {
        key: tab.name,
        path: tab.name,
        as: _components.__experimentalItem,
        isAction: true
      }, (0, _element.createElement)(_components.__experimentalHStack, {
        justify: "space-between"
      }, (0, _element.createElement)(_components.FlexItem, null, (0, _element.createElement)(_components.__experimentalTruncate, null, tab.title)), (0, _element.createElement)(_components.FlexItem, null, (0, _element.createElement)(_icons.Icon, {
        icon: (0, _i18n.isRTL)() ? _icons.chevronLeft : _icons.chevronRight
      }))));
    }))))), sections.length && sections.map(section => {
      return (0, _element.createElement)(_components.__experimentalNavigatorScreen, {
        key: `${section.name}-menu`,
        path: section.name
      }, (0, _element.createElement)(_components.Card, {
        isBorderless: true,
        size: "large"
      }, (0, _element.createElement)(_components.CardHeader, {
        isBorderless: false,
        justify: "left",
        size: "small",
        gap: "6"
      }, (0, _element.createElement)(_components.__experimentalNavigatorBackButton, {
        icon: (0, _i18n.isRTL)() ? _icons.chevronRight : _icons.chevronLeft,
        "aria-label": (0, _i18n.__)('Navigate to the previous view')
      }), (0, _element.createElement)(_components.__experimentalText, {
        size: "16"
      }, section.tabLabel)), (0, _element.createElement)(_components.CardBody, null, section.content)));
    }));
  }

  return modalContent;
}
