// @wordpress/format-library/build/code/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.code = void 0;

var _element = require("@wordpress/element");

var _i18n = require("@wordpress/i18n");

var _richText = require("@wordpress/rich-text");

var _blockEditor = require("@wordpress/block-editor");

var _icons = require("@wordpress/icons");

/**
 * WordPress dependencies
 */
const name = 'core/code';
const title = (0, _i18n.__)('Inline code');
const code = {
  name,
  title,
  tagName: 'code',
  className: null,

  __unstableInputRule(value) {
    const BACKTICK = '`';
    const {
      start,
      text
    } = value;
    const characterBefore = text.slice(start - 1, start); // Quick check the text for the necessary character.

    if (characterBefore !== BACKTICK) {
      return value;
    }

    const textBefore = text.slice(0, start - 1);
    const indexBefore = textBefore.lastIndexOf(BACKTICK);

    if (indexBefore === -1) {
      return value;
    }

    const startIndex = indexBefore;
    const endIndex = start - 2;

    if (startIndex === endIndex) {
      return value;
    }

    value = (0, _richText.remove)(value, startIndex, startIndex + 1);
    value = (0, _richText.remove)(value, endIndex, endIndex + 1);
    value = (0, _richText.applyFormat)(value, {
      type: name
    }, startIndex, endIndex);
    return value;
  },

  edit(_ref) {
    let {
      value,
      onChange,
      onFocus,
      isActive
    } = _ref;

    function onClick() {
      onChange((0, _richText.toggleFormat)(value, {
        type: name,
        title
      }));
      onFocus();
    }

    return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_blockEditor.RichTextShortcut, {
      type: "access",
      character: "x",
      onUse: onClick
    }), (0, _element.createElement)(_blockEditor.RichTextToolbarButton, {
      icon: _icons.code,
      title: title,
      onClick: onClick,
      isActive: isActive,
      role: "menuitemcheckbox"
    }));
  }

};
exports.code = code;
