// @wordpress/edit-post/build/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.initializeEditor = initializeEditor;
Object.defineProperty(exports, "store", {
  enumerable: true,
  get: function () {
    return _store.store;
  }
});

var _element = require("@wordpress/element");

require("@wordpress/core-data");

require("@wordpress/format-library");

var _data = require("@wordpress/data");

var _preferences = require("@wordpress/preferences");

var _store = require("./store");

var _editor = _interopRequireDefault(require("./editor"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

/**
 * Initializes the Editor and returns a componentProvider
 * that can be registered with `AppRegistry.registerComponent`
 *
 * @param {string} id       Unique identifier for editor instance.
 * @param {Object} postType Post type of the post to edit.
 * @param {Object} postId   ID of the post to edit (unused right now)
 */
function initializeEditor(id, postType, postId) {
  (0, _data.dispatch)(_preferences.store).setDefaults('core/edit-post', {
    editorMode: 'visual',
    fixedToolbar: false,
    fullscreenMode: true,
    hiddenBlockTypes: [],
    inactivePanels: [],
    isPublishSidebarEnabled: true,
    openPanels: ['post-status'],
    preferredStyleVariations: {},
    welcomeGuide: true
  });
  return (0, _element.createElement)(_editor.default, {
    postId: postId,
    postType: postType
  });
}
