// @wordpress/edit-post/build/components/header/template-title/template-description.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TemplateDescription;

var _element = require("@wordpress/element");

var _data = require("@wordpress/data");

var _components = require("@wordpress/components");

var _store = require("../../../store");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function TemplateDescription() {
  const {
    description,
    title
  } = (0, _data.useSelect)(select => {
    const {
      getEditedPostTemplate
    } = select(_store.store);
    return {
      title: getEditedPostTemplate().title,
      description: getEditedPostTemplate().description
    };
  }, []);

  if (!description) {
    return null;
  }

  return (0, _element.createElement)("div", {
    className: "edit-site-template-details__group"
  }, (0, _element.createElement)(_components.__experimentalHeading, {
    level: 4,
    weight: 600
  }, title), (0, _element.createElement)(_components.__experimentalText, {
    className: "edit-post-template-details__description",
    size: "body",
    as: "p",
    style: {
      marginTop: '12px'
    }
  }, description));
}
