// @wordpress/editor/build/components/post-taxonomies/most-used-terms.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = MostUsedTerms;

var _element = require("@wordpress/element");

var _lodash = require("lodash");

var _components = require("@wordpress/components");

var _data = require("@wordpress/data");

var _coreData = require("@wordpress/core-data");

var _terms2 = require("../../utils/terms");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
const MIN_MOST_USED_TERMS = 3;
const DEFAULT_QUERY = {
  per_page: 10,
  orderby: 'count',
  order: 'desc',
  hide_empty: true,
  _fields: 'id,name,count',
  context: 'view'
};

function MostUsedTerms(_ref) {
  let {
    onSelect,
    taxonomy
  } = _ref;
  const {
    _terms,
    showTerms
  } = (0, _data.useSelect)(select => {
    const mostUsedTerms = select(_coreData.store).getEntityRecords('taxonomy', taxonomy.slug, DEFAULT_QUERY);
    return {
      _terms: mostUsedTerms,
      showTerms: (mostUsedTerms === null || mostUsedTerms === void 0 ? void 0 : mostUsedTerms.length) >= MIN_MOST_USED_TERMS
    };
  }, []);

  if (!showTerms) {
    return null;
  }

  const terms = (0, _terms2.unescapeTerms)(_terms);
  const label = (0, _lodash.get)(taxonomy, ['labels', 'most_used']);
  return (0, _element.createElement)("div", {
    className: "editor-post-taxonomies__flat-term-most-used"
  }, (0, _element.createElement)(_components.BaseControl.VisualLabel, {
    as: "h3",
    className: "editor-post-taxonomies__flat-term-most-used-label"
  }, label), (0, _element.createElement)("ul", {
    role: "list",
    className: "editor-post-taxonomies__flat-term-most-used-list"
  }, terms.map(term => (0, _element.createElement)("li", {
    key: term.id
  }, (0, _element.createElement)(_components.Button, {
    variant: "link",
    onClick: () => onSelect(term)
  }, term.name)))));
}
