// @wordpress/edit-post/build/components/sidebar/post-url/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PostURL;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _editor = require("@wordpress/editor");

/**
 * WordPress dependencies
 */
function PostURL() {
  // Use internal state instead of a ref to make sure that the component
  // re-renders when the popover's anchor updates.
  const [popoverAnchor, setPopoverAnchor] = (0, _element.useState)(null); // Memoize popoverProps to avoid returning a new object every time.

  const popoverProps = (0, _element.useMemo)(() => ({
    anchor: popoverAnchor
  }), [popoverAnchor]);
  return (0, _element.createElement)(_editor.PostURLCheck, null, (0, _element.createElement)(_components.PanelRow, {
    className: "edit-post-post-url",
    ref: setPopoverAnchor
  }, (0, _element.createElement)("span", null, (0, _i18n.__)('URL')), (0, _element.createElement)(_components.Dropdown, {
    popoverProps: popoverProps,
    position: "bottom left",
    className: "edit-post-post-url__dropdown",
    contentClassName: "edit-post-post-url__dialog",
    focusOnMount: true,
    renderToggle: _ref => {
      let {
        isOpen,
        onToggle
      } = _ref;
      return (0, _element.createElement)(PostURLToggle, {
        isOpen: isOpen,
        onClick: onToggle
      });
    },
    renderContent: _ref2 => {
      let {
        onClose
      } = _ref2;
      return (0, _element.createElement)(_editor.PostURL, {
        onClose: onClose
      });
    }
  })));
}

function PostURLToggle(_ref3) {
  let {
    isOpen,
    onClick
  } = _ref3;
  const label = (0, _editor.usePostURLLabel)();
  return (0, _element.createElement)(_components.Button, {
    className: "edit-post-post-url__toggle",
    variant: "tertiary",
    "aria-expanded": isOpen // translators: %s: Current post URL.
    ,
    "aria-label": (0, _i18n.sprintf)((0, _i18n.__)('Change URL: %s'), label),
    onClick: onClick
  }, label);
}
