// @wordpress/editor/build-module/components/post-excerpt/index.js
import { createElement } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { ExternalLink, TextareaControl } from '@wordpress/components';
import { withSelect, withDispatch } from '@wordpress/data';
import { compose } from '@wordpress/compose';
/**
 * Internal dependencies
 */

import { store as editorStore } from '../../store';

function PostExcerpt(_ref) {
  let {
    excerpt,
    onUpdateExcerpt
  } = _ref;
  return createElement("div", {
    className: "editor-post-excerpt"
  }, createElement(TextareaControl, {
    __nextHasNoMarginBottom: true,
    label: __('Write an excerpt (optional)'),
    className: "editor-post-excerpt__textarea",
    onChange: value => onUpdateExcerpt(value),
    value: excerpt
  }), createElement(ExternalLink, {
    href: __('https://wordpress.org/support/article/settings-sidebar/#excerpt')
  }, __('Learn more about manual excerpts')));
}

export default compose([withSelect(select => {
  return {
    excerpt: select(editorStore).getEditedPostAttribute('excerpt')
  };
}), withDispatch(dispatch => ({
  onUpdateExcerpt(excerpt) {
    dispatch(editorStore).editPost({
      excerpt
    });
  }

}))])(PostExcerpt);
