// @wordpress/block-editor/build/components/inspector-controls/groups.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _components = require("@wordpress/components");

/**
 * WordPress dependencies
 */
const InspectorControlsDefault = (0, _components.createSlotFill)('InspectorControls');
const InspectorControlsAdvanced = (0, _components.createSlotFill)('InspectorAdvancedControls');
const InspectorControlsBorder = (0, _components.createSlotFill)('InspectorControlsBorder');
const InspectorControlsColor = (0, _components.createSlotFill)('InspectorControlsColor');
const InspectorControlsDimensions = (0, _components.createSlotFill)('InspectorControlsDimensions');
const InspectorControlsTypography = (0, _components.createSlotFill)('InspectorControlsTypography');
const InspectorControlsListView = (0, _components.createSlotFill)('InspectorControlsListView');
const groups = {
  default: InspectorControlsDefault,
  advanced: InspectorControlsAdvanced,
  border: InspectorControlsBorder,
  color: InspectorControlsColor,
  dimensions: InspectorControlsDimensions,
  list: InspectorControlsListView,
  typography: InspectorControlsTypography
};
var _default = groups;
exports.default = _default;
