// @wordpress/edit-post/build/components/sidebar/post-template/create-modal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PostTemplateCreateModal;

var _element = require("@wordpress/element");

var _data = require("@wordpress/data");

var _editor = require("@wordpress/editor");

var _blocks = require("@wordpress/blocks");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _url = require("@wordpress/url");

var _store = require("../../../store");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
const DEFAULT_TITLE = (0, _i18n.__)('Custom Template');

function PostTemplateCreateModal(_ref) {
  let {
    onClose
  } = _ref;
  const defaultBlockTemplate = (0, _data.useSelect)(select => select(_editor.store).getEditorSettings().defaultBlockTemplate, []);
  const {
    __unstableCreateTemplate,
    __unstableSwitchToTemplateMode
  } = (0, _data.useDispatch)(_store.store);
  const [title, setTitle] = (0, _element.useState)('');
  const [isBusy, setIsBusy] = (0, _element.useState)(false);

  const cancel = () => {
    setTitle('');
    onClose();
  };

  const submit = async event => {
    event.preventDefault();

    if (isBusy) {
      return;
    }

    setIsBusy(true);
    const newTemplateContent = defaultBlockTemplate !== null && defaultBlockTemplate !== void 0 ? defaultBlockTemplate : (0, _blocks.serialize)([(0, _blocks.createBlock)('core/group', {
      tagName: 'header',
      layout: {
        inherit: true
      }
    }, [(0, _blocks.createBlock)('core/site-title'), (0, _blocks.createBlock)('core/site-tagline')]), (0, _blocks.createBlock)('core/separator'), (0, _blocks.createBlock)('core/group', {
      tagName: 'main'
    }, [(0, _blocks.createBlock)('core/group', {
      layout: {
        inherit: true
      }
    }, [(0, _blocks.createBlock)('core/post-title')]), (0, _blocks.createBlock)('core/post-content', {
      layout: {
        inherit: true
      }
    })])]);
    await __unstableCreateTemplate({
      slug: (0, _url.cleanForSlug)(title || DEFAULT_TITLE),
      content: newTemplateContent,
      title: title || DEFAULT_TITLE
    });
    setIsBusy(false);
    cancel();

    __unstableSwitchToTemplateMode(true);
  };

  return (0, _element.createElement)(_components.Modal, {
    title: (0, _i18n.__)('Create custom template'),
    closeLabel: (0, _i18n.__)('Close'),
    onRequestClose: cancel,
    className: "edit-post-post-template__create-modal"
  }, (0, _element.createElement)("form", {
    className: "edit-post-post-template__create-form",
    onSubmit: submit
  }, (0, _element.createElement)(_components.Flex, {
    align: "flex-start",
    gap: 8
  }, (0, _element.createElement)(_components.FlexItem, null, (0, _element.createElement)(_components.TextControl, {
    label: (0, _i18n.__)('Name'),
    value: title,
    onChange: setTitle,
    placeholder: DEFAULT_TITLE,
    disabled: isBusy,
    help: (0, _i18n.__)('Describe the template, e.g. "Post with sidebar". Custom templates can be applied to any post or page.')
  }))), (0, _element.createElement)(_components.Flex, {
    justify: "flex-end",
    expanded: false
  }, (0, _element.createElement)(_components.FlexItem, null, (0, _element.createElement)(_components.Button, {
    variant: "tertiary",
    onClick: cancel
  }, (0, _i18n.__)('Cancel'))), (0, _element.createElement)(_components.FlexItem, null, (0, _element.createElement)(_components.Button, {
    variant: "primary",
    type: "submit",
    isBusy: isBusy,
    "aria-disabled": isBusy
  }, (0, _i18n.__)('Create'))))));
}
